native-addons-are-close-to-being-on-par-with-js-modules.md

```diff
@@ -0,0 +1,113 @@
+
+These differences have created a sizeable maintenance burden for those who choose to publish native addons, and for those whose deployed applications include native addons.
+
+### 1. Compilation
+This will always be a fundamental difference between native addons and JavaScript modules. The process whereby source code written in C or C++ (or even a different language such as Rust or Go) gets bundled into a native addon will always be an extra step in addition to publishing an npm package which users can consume. In fact, if the maintainer does not take the time to provide pre-built packages by using tools such as [node-pre-gyp][] or [prebuild][], then those who would depend on the native addon have to install a compiler and build tool chain to allow the native addon package to build on their system. Nevertheless, npm packages providing pre-built native addons can make it appear as though a native addon can be installed as easily as a JavaScript module.
+
+### 2. Platform And Architecture Independence
+A single `.js` file defines a single JavaScript module because fundamentally it is nothing more than plain text which is read in from the file system by Node.js and then interpreted. In contrast, a native addon is a binary containing machine code much like Node.js itself. Thus, a single binary can only serve a specific type of machine running a specific type of operating system. For example, a single binary can only serve Windows running on x86 (Intel or AMD). A different binary is needed for OSX on x86, and yet another one is needed for Linux on x86 and so on. Addon maintainers' tools such as node-pre-gyp and prebuild can bridge this gap as well. They make it possible for maintainers to produce a binary for each platform/architecture combination they wish to support so that at package installation time, the npm install script can choose such a binary, download it, and give it to Node.js for loading and execution.
+
+### 3. Node.js Version Independence
+JavaScript modules, once written, will work on the version of Node.js for which they were written, *and on all subsequent versions*. In contrast, native addons are tied to the version of Node.js for which they were written. For a given platform and architecture, a different binary has to be provided for version 8.x of Node.js than for version 10.x. The reason for this is that the native API provided by Node.js to native addon developers can change from one major version to the next in such a way that the native addon will no longer load, or, in the worst case, it will load but crash seemingly inexplicably. To save many hours of hunting down seemingly inexplicable crashes, a simple versioning scheme based on a single integer value was added, whereby the Node.js version against which an addon was built is recorded within the addon and examined upon loading the addon. If the value does not match the value carried by Node.js, it will fail to load the addon, producing a familiar message:
+
```

Comment:
"familiar message" might not apply to persons who did not work with these issues in the past.

Reply:
Fixed ✓